### PR TITLE
Add JVM args config

### LINKED ISSUE #695: Add support for JVM args for Scala subprocess-based kernel
Data scientists like to have it their way, e.g. `-Xmx50G` ;)

### PR DESCRIPTION
**Pull request by jeremyrsmith**
(Includes #1015)

- Add config for JVM args (fixes #695)
- Add UI for Scala version and JVM args
- Don't inherit the whole classpath of server process (unless `POLYNOTE_INHERIT_CLASSPATH` env is set, or listing the required distribution JARs fails – to allow running from IDE)

### CODE REVIEW

File: polynote-frontend/polynote/ui/component/notebook/notebookconfig.ts [hunk 1]
```diff
@@ -401,15 +401,33 @@ class SparkConf {
...
+            "2.12": "2.12"
+        }
 
-    constructor(envHandler: StateView<Record<string, string> | undefined>) {
         this.el = div(['notebook-env', 'notebook-config-section'], [
-            h3([], ['Environment Variables']),
-            para([], ['Set environment variables here. Please note this is only supported when kernels are launched as a subprocess (default).']),
-            this.container = div(['env-list'], [])
+            h3([], ['Kernel configuration']),
+            para([], ['Please note this is only supported when kernels are launched as a subprocess (default).']),
```
Comment (jonathanindig) [inline polynote-frontend/polynote/ui/component/notebook/notebookconfig.ts:423]: Consider changing this wording since it's no longer possible to launch kernels in-process.

File: polynote-frontend/polynote/ui/component/notebook/notebookconfig.ts [hunk 2]
```diff
@@ -401,15 +401,33 @@ class SparkConf {
...
     private container: TagElement<"div">;
+    private jvmArgsInput: TagElement<"input">;
+    private scalaVersionInput: DropdownElement;
+
+    constructor(configState: StateView<NotebookConfig>) {
+        const envHandler = configState.view("env");
+        const jvmArgsHandler = configState.view("jvmArgs");
+        const scalaVersionHandler = configState.view("scalaVersion");
+
+        // TODO: this could come from the server
```
Comment (jonathanindig) [inline polynote-frontend/polynote/ui/component/notebook/notebookconfig.ts:415]: +1, seems like a perfect candidate for `ServerHandshake`.

File: polynote-frontend/polynote/ui/component/notebook/notebookconfig.ts [hunk 3]
```diff
@@ -401,15 +401,33 @@ class SparkConf {
...
-            this.container = div(['env-list'], [])
+            h3([], ['Kernel configuration']),
+            para([], ['Please note this is only supported when kernels are launched as a subprocess (default).']),
+            h4([], 'Scala version:'),
+            para([], `If using Spark, the Scala version must match that of your Spark installation. "Default" will use Polynote's configured Scala version, or auto-detect the appropriate version.`),
+            this.scalaVersionInput = dropdown(['scala-version'], {"": "Default", ...availableScalaVersions}, scalaVersionHandler.state || ""),
+            h4([], 'Environment variables:'),
+            this.container = div(['env-list'], []),
+            para([], ['Additional JVM arguments:']),
+            this.jvmArgsInput = textbox(['jvm-args'], "JVM Arguments", jvmArgsHandler.state).attr("size", "64")
```
Comment (jonathanindig) [inline polynote-frontend/polynote/ui/component/notebook/notebookconfig.ts:430]: Might be good to specify that they should be in the command line format, e.g., `-Dfoo=bar`

File: polynote-frontend/polynote/ui/component/notebook/notebookconfig.ts (hunk 1 above)
Comment (jeremyrsmith) [reply polynote-frontend/polynote/ui/component/notebook/notebookconfig.ts:423]: It is possible – if you're not using Spark and have `behavior.kernel_isolation = spark`.

File: polynote-frontend/polynote/ui/component/notebook/notebookconfig.ts [hunk 4]
```diff
@@ -401,15 +401,33 @@ class SparkConf {
...
 
-    constructor(envHandler: StateView<Record<string, string> | undefined>) {
         this.el = div(['notebook-env', 'notebook-config-section'], [
-            h3([], ['Environment Variables']),
-            para([], ['Set environment variables here. Please note this is only supported when kernels are launched as a subprocess (default).']),
-            this.container = div(['env-list'], [])
+            h3([], ['Kernel configuration']),
+            para([], ['Please note this is only supported when kernels are launched as a subprocess (default).']),
+            h4([], 'Scala version:'),
+            para([], `If using Spark, the Scala version must match that of your Spark installation. "Default" will use Polynote's configured Scala version, or auto-detect the appropriate version.`),
```
[the remark this answers is not in the document]
Comment (jeremyrsmith) [reply polynote-frontend/polynote/ui/component/notebook/notebookconfig.ts:425]: * If an explicit (not "Default") Scala version is configured in the notebook, it will use that.
* Otherwise, if polynote is configured with `kernel.scala_version` then it will use that.
* Otherwise, if you're launching a Spark kernel, it will try to detect which Scala version your Spark installation wants.
* Otherwise, it will use the same Scala version that the running server was built with.

File: polynote-frontend/polynote/ui/component/notebook/notebookconfig.ts (hunk 3 above)
Comment (jeremyrsmith) [reply polynote-frontend/polynote/ui/component/notebook/notebookconfig.ts:431]: Good point – also it should be an `h4`. Fixed.

File: polynote-kernel/src/main/scala/polynote/kernel/remote/transport.scala [hunk 5]
```diff
@@ -338,22 +339,33 @@ object SocketTransport {
 
     private def listJars(path: Path): RIO[BaseEnv, Seq[Path]] = listFiles(path)
       .map(_.filter(_.getFileName.toString.endsWith(".jar")))
-      .catchSome {
-        case NonFatal(err) => Logging.error(s"Failed to list JARs in $path", err).as(Seq.empty)
+      .tapError {
```
Comment (jonathanindig) [inline polynote-kernel/src/main/scala/polynote/kernel/remote/transport.scala:342]: Why the change from `catchSome` -> `tapError` ? With this change the error will bubble up, right?

Comment (jeremyrsmith) [reply polynote-kernel/src/main/scala/polynote/kernel/remote/transport.scala:342]: Yes, I want it to bubble up so I can fall back to inheriting the classpath of the server (so things can run from IDE or sbt without having to build the distribution first)

File: polynote-frontend/polynote/ui/component/notebook/notebookconfig.ts (hunk 1 above)
Comment (jonathanindig) [reply polynote-frontend/polynote/ui/component/notebook/notebookconfig.ts:423]: Oh, I guess I misunderstood the changes in the other PR. Thanks for the clarification.

File: polynote-frontend/polynote/ui/component/notebook/notebookconfig.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jonathanindig) [reply polynote-frontend/polynote/ui/component/notebook/notebookconfig.ts:425]: In that case, maybe this would be more in line with the order of operations you're describing?
```suggestion
            para([], `If using Spark, the Scala version must match that of your Spark installation. "Default" will auto-detect the appropriate version if possible, otherwise falling back to using Polynote's configured Scala version`),
```

File: polynote-kernel/src/main/scala/polynote/kernel/remote/transport.scala (hunk 5 above)
Comment (jonathanindig) [reply polynote-kernel/src/main/scala/polynote/kernel/remote/transport.scala:342]: Ah, so where is the error being caught now?

File: polynote-frontend/polynote/ui/component/notebook/notebookconfig.ts [hunk 6]
```diff
@@ -401,15 +401,34 @@ class SparkConf {
...
+            this.scalaVersionInput = dropdown(['scala-version'], {"": "Default", ...availableScalaVersions}, scalaVersionHandler.state || ""),
+            h4([], 'Environment variables:'),
+            this.container = div(['env-list'], []),
+            h4([], ['Additional JVM arguments:']),
+            para([], ['Extra command-line arguments to the JVM, e.g. ', tag('code', [], {}, '"-Dmy.prop=a value" -Xmx200m')]),
```
Comment (jonathanindig) [inline polynote-frontend/polynote/ui/component/notebook/notebookconfig.ts:430]: `Xmx200m` might be a dangerous example 😅

File: polynote-kernel/src/main/scala/polynote/kernel/remote/transport.scala [hunk 7]
```diff
@@ -401,19 +413,61 @@ object SocketTransport {
...
-          "--kernelFactory" :: classTag[KernelFactory].runtimeClass.getName ::
-          Nil
+      private def findJava: URIO[BaseEnv, String] =
+        property("java.home").mapError(_.getMessage).someOrFail("No java.home property is set")
+          .map(home => Paths.get(home, "bin", "java").toString)
+          .tapError(err => Logging.warn("Couldn't find java executable; will just use 'java' ($err)"))
+          .orElse(ZIO.succeed("java"))
+
+      // parse a JVM args string into a list of args
+      private def parseJVMArgs(str: String): List[String] = {
```
Comment (jonathanindig) [inline polynote-kernel/src/main/scala/polynote/kernel/remote/transport.scala:423]: Is this parsing just turning `"-Dfoo=bar -Dbar=baz -Xmx200m"` into `List("-Dfoo=bar",  "-Dbar=baz",  "-Xmx200m)"` or is there more to it?

Comment (jonathanindig) [reply polynote-kernel/src/main/scala/polynote/kernel/remote/transport.scala:423]: Also, what do you think about adding some tests to this parser? 😉 

idk, I see strings and regexes and I think "tests" 😛

**Comment by jonathanindig** [comment]
Does this work with non-Spark kernels? I only found a change to `DeploySparkSubmit.scala`.

**Comment by jonathanindig** [comment]
Hmm, I think it might be a bit nicer for users to specify the JVM args as a `List[String]` - i.e., multiple inputs rather than a single one. Any reason it's specified as a plain `String` in this implementation?

**Comment by jeremyrsmith** [comment]
@jonathanindig yes, there are also changes to `DeployJava` – it's just in `transport.scala` rather than its own file.

**Comment by jeremyrsmith** [comment]
@jonathanindig I thought about making it a `List[String]`. I guess there's no reason it couldn't be; it just seemed more complicated for little benefit (that I could see).

**Comment by jonathanindig** [comment]
Also, another thought - is there any sanitization we should be doing to the jvm args for security purposes. Or does `ProcessBuilder` handle all that for us?

**Comment by jonathanindig** [comment]
> @jonathanindig yes, there are also changes to `DeployJava` – it's just in `transport.scala` rather than its own file.

Oh, of course! 🤦

**Comment by jonathanindig** [comment]
> @jonathanindig I thought about making it a `List[String]`. I guess there's no reason it couldn't be; it just seemed more complicated for little benefit (that I could see).

I'm thinking that multiple inputs in the UI would be easier for users to input than a single large string which is a bit of a pain to write in the config.

**Comment by jeremyrsmith** [comment]
@jonathanindig I don't think there's all that much we could do to sanitize or validate the args – short of completely modeling the args we want to support (i.e. separating out system properties, the various memory flags, GC configuration, debugger agent, etc.) Maybe that would be a better way to go? It would help protect against foot-gunning yourself with a wrong JVM flag, I suppose (at the expense of a lot of complexity).

But if you're worried about injecting additional commands, I don't think it's possible with ProcessBuilder. E.g.

```scala
import sys.process._
Seq("echo", "hello", "&&", "echo", "goodbye").!
// result: hello && echo goodbye
```

**Comment by jeremyrsmith** [comment]
@jonathanindig

> I'm thinking that multiple inputs in the UI would be easier for users to input than a single large string which is a bit of a pain to write in the config.

I'm not sure I agree... usually these things come from copypasta, so having a single field to receive copypasta is "easier" in the common case (IMHO)

**Comment by jeremyrsmith** [comment]
Let's ask @kamilkloch, who created #695 😄 

@kamilkloch, for configuring JVM options to the kernel process, would you rather have a list of args with the +/- UI (like dependencies has) or a single field where you can add or paste a bunch of JVM flags (but you'll have to quote any flags that include a space)?

**Comment by kamilkloch** [comment]
@jeremyrsmith @jonathanindig First, thanks for solving #695, and thanks for asking. I gravitate towards a single field (and all flags neatly in one place) but honestly, both solutions are good :)

**Review by jonathanindig — APPROVED** [review]
